### PR TITLE
fix(cli): dispatch init to the setup wizard (#181)

### LINKED ISSUE #181: Dispatch init in cli.ts so the short npx form reaches the setup wizard
## Description

`npx blamechris/repo-relay init` runs the `repo-relay` bin (`dist/cli.js`) because npm selects the bin matching the package name; `cli.js` ignores the `init` argument and exits 1 wanting `DISCORD_BOT_TOKEN`, so the short form never reaches the setup wizard. PR #180 switched the README to the longer `npx -p github:blamechris/repo-relay repo-relay-init` form as a docs-only workaround; this issue is the real fix so the short form works again.

## Acceptance Criteria

- [ ] `src/cli.ts` dispatches `init`: if `process.argv[2] === 'init'`, run the setup wizard before any env-var checks
- [ ] `dist/` rebuilt so `npx blamechris/repo-relay init` reaches the wizard

---
Filed from review of PR #180 (2026-08-05 portfolio launch).

### PR DESCRIPTION
**Pull request by blamechris**
Closes #181

## Problem

`npx blamechris/repo-relay init` runs the `repo-relay` bin (`dist/cli.js`) because npm selects the bin matching the package name; `cli.js` ignored the argument and exited 1 wanting `DISCORD_BOT_TOKEN`, so the short form never reached the setup wizard. #180 worked around it in docs with the longer `npx -p github:blamechris/repo-relay repo-relay-init` form.

## Change

- **`src/cli.ts`** — `process.argv[2] === 'init'` dispatches to the wizard before any env-var checks, via dynamic import so the GitHub Actions hot path doesn't load `prompts`/`execSync`.
- **`src/setup.ts`** — `main()` becomes exported `runSetup()` behind a realpath entry-point guard (same pattern as `cli.ts`), so importing `setup.js` cannot auto-run the wizard while the `repo-relay-init` bin keeps working unchanged.
- **`src/__tests__/cli-init-dispatch.test.ts`** — spawns the committed `dist/cli.js`: `init` reaches the wizard with no env set; the bare Actions path still demands `DISCORD_BOT_TOKEN`; importing `setup.js` does not start the wizard.
- **README** — Quick Start returns to the short form.
- **`dist/`** rebuilt (dist-freshness gate). Note: `dist/cli.js` and `dist/setup.js` also flip to mode 755 — npm's bin-linking sets the exec bit on shebang'd bin targets; correct and kept.

## Verification

- Full suite: 403 tests green; typecheck + lint clean.
- End-to-end: packed the worktree, installed it into a scratch project — `npx repo-relay init` reaches the wizard; `npx repo-relay-init` unchanged; bare `node dist/cli.js` still env-gates.
- Surfaced in testing: the wizard exits 0 when stdin hits EOF mid-prompt — pre-existing on both bins, filed as #183 (the new test deliberately doesn't pin that exit code).

The wizard's Step 1 privileged-intents copy is intentionally untouched here — tracked as #182, next up.

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #184

### Summary

Makes `npx blamechris/repo-relay init` reach the setup wizard by dispatching `process.argv[2] === 'init'` from `cli.ts` before any env-var checks, and converts `setup.ts`'s bare `main()` call into an exported `runSetup()` behind a realpath entry-point guard so importing the module can't start the wizard. README's Quick Start reverts from the #180 docs workaround to the short form, and `dist/` is rebuilt.

This hits both acceptance criteria on #181 exactly, and adds the entry-point guard and spawn-based tests the issue didn't ask for. I rebuilt the head commit (`db8de63`) locally: `tsc` is clean, `eslint` is clean, `git status --porcelain dist/` is empty after the build, and the three new tests pass in 534 ms.

### Strengths

- **Guard pattern matches the existing one exactly.** `setup.ts:296-303`'s `isEntryPoint()` is character-identical to the one already at the bottom of `cli.ts` — `process.argv[1]` null-check, realpath both sides, `catch { return false }`. No new idiom introduced for a second copy of the same problem.
- **Actions hot path is genuinely behavior-identical.** `action.yml` runs `node dist/cli.js` with no arguments, so `argv[2]` is `undefined`, the dispatch is skipped, and `[repo-relay] Starting...` is still the first line emitted. Verified by spawning the built artifact with an empty env.
- **Dispatch placement is correct.** It sits ahead of all three env-var checks and the channel-config load, so the wizard is reachable with a completely empty environment — verified with `env -i`.
- **Dynamic import earns its keep.** `prompts` and `child_process.execSync` stay out of the Actions path, and `prompts` is already a prod dependency so `npm ci --omit=dev` in `action.yml` still resolves it on the init path.
- **`dist/` is genuinely fresh, including the mode flip.** The 644→755 change on `dist/cli.js` and `dist/setup.js` is stable under the freshness gate: `tsc` overwrites in place and preserves the mode, so a CI rebuild produces no diff. I confirmed this rather than assuming it.
- **Test comments explain the non-obvious choices** — why there's no exit-code assertion on the init case, why signal kills reject instead of resolving, why the tests target `dist/` rather than `src/`.
- **Honest scoping.** The wizard's exit-0-on-EOF quirk was found during this work, filed as #183, and deliberately left unpinned by the test instead of being encoded as expected behavior.

### Issues Found

#### Critical

None.

#### Suggestions

| File:Line | Issue | Suggested Fix |
|---|---|---|
| `src/cli.ts:29-33` | The dispatch runs inside `main()`, which executes only after `cli.ts`'s hoisted static imports have already evaluated `./index.js` → `discord.js` and `./db/state.js` → `better-sqlite3`. So `npx ... init` now loads the entire Discord + native-SQLite stack before printing a single wizard line, and a `better-sqlite3` load failure (no prebuild for the user's platform/Node, no toolchain to build one) crashes the wizard before Step 1 — a failure mode the standalone `repo-relay-init` bin never had. The inline comment is accurate about `prompts`/`execSync` but describes only one direction, which could mislead a future reader into thinking `init` is a light path. Measured locally: ~240 ms to first prompt via `cli.js init` vs ~100 ms via `setup.js` — the latency is fine; the robustness is the real point. | Either move `RepoRelay`/`getChannelConfig`/`shouldSkipEvent` to a dynamic `import()` after the dispatch (the handler payload imports are already `import type`, so they erase), or extend the comment to state that `init` still pays the full runtime load. |
| `src/__tests__/cli-init-dispatch.test.ts` (whole file) | Nothing covers the **positive** case of the new guard: that `node dist/setup.js` still starts the wizard. Test 1 reaches `runSetup` through `cli.js`'s import, and test 3 asserts that importing does *not* run it — so if `isEntryPoint()` ever started returning `false` (argv[1] shape change, a bundler, a realpath edge case), the documented `repo-relay-init` bin would silently do nothing and the suite would stay green. This is the one behavior the PR changed that has no regression net under it. I confirmed manually that it works today. | Add a fourth case: `const { stdout } = await run([SETUP_PATH]); expect(stdout).toContain('repo-relay Setup');` |
| `src/__tests__/cli-init-dispatch.test.ts:32` | The 30 s `execFile` timeout can never fire — `vitest.config.ts` sets no `testTimeout`, so vitest's 5 s default kills the test first. That makes the deliberately-written "signal kills surface as `code:null` + signal, fail loudly" branch at line 41 unreachable in practice, and a genuine hang reports as a generic vitest timeout instead of the specific error. Runtimes are 242/147/144 ms locally so there's real headroom, but the slowest realistic case is a cold `ubuntu-latest` runner loading `discord.js` plus the `better-sqlite3` binding. | Give the three cases an explicit timeout (`it('...', async () => { ... }, 30_000)`) so the `execFile` guard is the one that actually fires — or drop the `execFile` timeout below 5 s so the intended error path is reachable. |

#### Nitpicks

| File:Line | Issue | Suggested Fix |
|---|---|---|
| `src/cli.ts:29` | Any `argv[2]` other than exactly `init` — `--help`, `-h`, `setup`, `Init` — falls through to the Actions path and prints `[repo-relay] ERROR: DISCORD_BOT_TOKEN is required`. That is precisely the confusing failure #181 was filed about, now one typo away rather than fixed as a class. Verified: `node dist/cli.js --help` prints the token error. | Optional `else if (process.argv[2])` printing a two-line usage hint (`repo-relay init` for setup; no args for the Actions runner) before exiting 1. |
| `src/__tests__/cli-init-dispatch.test.ts:34` | `env: { PATH: process.env.PATH }` drops `SystemRoot`/`COMSPEC`, so these spawns fail on a Windows dev machine. CI is `ubuntu-latest` only, so nothing is gated — it's purely a contributor-debugging cost. | A one-line comment noting the tests are POSIX-only, or spread `SystemRoot` back in when `process.platform === 'win32'`. |
| `src/setup.ts:308` vs `src/cli.ts:220` | A throw inside `runSetup()` surfaces as `Error: <msg>` via the `repo-relay-init` bin but as `[repo-relay] Unhandled error: <msg>` via `cli.js init`. Same exit code (1), only the prefix differs — worth knowing when triaging a user's pasted output. | Leave as-is, or prefix `setup.ts`'s handler with `[repo-relay]` for consistency. |

### Architecture Notes

- **Entry-point guard consistency:** correct — this is now the same guard in both executables rather than a second approach. Worth keeping in mind that any future change to the pattern has two call sites.
- **No import cycle:** `setup.ts` imports only `./utils/errors.js` and `./setup/workflow-template.js`; nothing imports `cli.ts`. The dynamic `import('./setup.js')` is a clean one-way edge.
- **Public API untouched:** `src/index.ts` does not re-export `setup.ts`, so `runSetup` is an internal seam, not a new supported surface — the right call.
- **Coverage config unperturbed:** `setup.ts` remains in `vitest.config.ts`'s coverage `exclude` list, so the `main()` → `runSetup()` rename doesn't disturb the ratchets. (CI runs `npm test`, not `npm run test:coverage`, so the thresholds aren't a merge gate here either way.)
- **One structural asymmetry worth naming:** `cli.js` is now both the unattended Actions runner and the interactive front door. Those two paths have very different weight and very different failure tolerances, which is what makes Suggestion 1 more than cosmetic.

### Test Coverage

Three subprocess tests against the committed `dist/`, which is what consumers actually execute — the right target, and the file comment says so explicitly. They cover: `init` reaches the wizard with an empty env and never prints `[repo-relay] Starting...`; the bare path still demands `DISCORD_BOT_TOKEN` and exits 1; importing `setup.js` prints `IMPORT_OK` without starting the wizard.

Flakiness and portability look sound for `ubuntu-latest`: `mkdtempSync` per run with `afterAll` cleanup, no fixed ports or sleeps, no reliance on repo state, and `child.stdin.end()` after spawn is race-free because the child simply blocks until EOF arrives. Deliberately omitting the exit-code assertion on the init case is the correct call given #183 — pinning `0` there would encode a bug as a contract.

Two gaps, both above: no positive-case test for the entry guard, and the 30 s `execFile` timeout is shadowed by vitest's 5 s default.

One structural note rather than a defect: because the tests assert against `dist/`, a `src`-only regression that isn't rebuilt leaves this suite green and is caught only by the separate `dist-freshness` job. That's the right trade-off for testing what ships, but it does mean these tests and that job are load-bearing together.

### Verdict

**Approved.** No blocking issues. The fix is minimal, correctly ordered ahead of the env-var checks, consistent with the existing entry-point guard, verified end-to-end, and `dist/` is genuinely fresh — I rebuilt it and the working tree stayed clean. The three suggestions are follow-on quality items, not merge blockers; the entry-guard positive-case test is the one I'd most like to see land, whether here or as a follow-up.

**Comment by blamechris** [comment]
Review round addressed in `e6889df` (+ `ad33126` for the regenerated declaration map):

- **Suggestion 2 (guard positive case):** added — `node dist/setup.js` as entry point must start the wizard; suite is now 404 tests.
- **Suggestion 3 (unreachable execFile timeout):** all four spawn tests now carry explicit 35s it-timeouts so the 30s execFile timeout (and the fail-loudly signal branch) can actually fire before vitest's 5s default kills the child.
- **Suggestion 1 (eager heavy imports on the init path):** deferred to #185 — restructuring cli.ts's static imports is beyond this fix's scope; the dispatch comment now documents the accepted trade-off.
- **Nitpick 1 (unknown-arg fallthrough):** a stray argument now gets a hint pointing at `init` before the env-var errors.
- **Nitpick 2 (PATH-only env on Windows):** documented in the test helper; CI is ubuntu-only, matching cli-best-effort.test.ts.
- **Nitpick 3 (error-prefix divergence):** wizard entry path now uses the `[repo-relay]` prefix.

File: src/__tests__/cli-init-dispatch.test.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        // Signal kills (the 30s timeout) surface as code:null + signal, and
+        // spawn failures as a string code — both must fail loudly rather
+        // than masquerade as an exit code
+        if (error && (error.signal || typeof error.code !== 'number')) {
+          return reject(error);
+        }
+        resolve({ code: (error?.code as number) ?? 0, stdout, stderr });
+      }
+    );
+    // Closed stdin = EOF at the first prompt → wizard cancels cleanly
```
[the remark this answers is not in the document]
Comment (blamechris) [reply src/__tests__/cli-init-dispatch.test.ts]: Right — the comment contradicted the behavior the test documents. Rewritten in `acea39e`: it now says EOF makes the event loop drain and node exit 0 mid-prompt (#183), and that closing stdin is there to prevent a hang, not to cancel cleanly.
